### PR TITLE
feat: simple merger

### PR DESCRIPTION
**Pull request by florian-hoenicke**
we need a containerized version of the merger to have it running on k8s

### CODE REVIEW

File: jinahub/indexers/merger/SimpleMerger/tests/test_simple_merger.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+__copyright__ = "Copyright (c) 2021 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
```
Comment (JoanFM) [inline jinahub/indexers/merger/SimpleMerger/tests/test_simple_merger.py:2]: lets have a proper split of unit and integration tests

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/indexers/merger/SimpleMerger/tests/unit/test_simple_merger.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                )
+                for j in range(3)
+            ])
+        for i in range(2)
+    ]
+
+
+def test_simple_merger(docs_matrix):
+    executor = SimpleMerger(default_traversal_paths=('r',))
+    executor.merge(docs_matrix=docs_matrix)
```
Comment (JoanFM) [inline jinahub/indexers/merger/SimpleMerger/tests/unit/test_simple_merger.py:37]: something should be asserted

**Review by JoanFM — COMMENTED** [review]
An integration test to prove how a Flow works would be needed as well. What do u think @tadejsv ?

File: jinahub/indexers/merger/SimpleMerger/simple_merger.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    The SimpleMerger merges the results of shards by appending all matches..
+
+    :param default_traversal_paths: traverse path on docs, e.g. ['r'], ['c']
+    :param args: additional arguments
+    :param kwargs: additional key value arguments
+    """
+
+    def __init__(self, default_traversal_paths: Tuple[str] = ('r',), **kwargs):
+
+        self.default_traversal_paths = default_traversal_paths
```
Comment (JoanFM) [inline jinahub/indexers/merger/SimpleMerger/simple_merger.py:21]: why putting it before `super().__init__(**kwargs). It's the same but looks weird

File: jinahub/indexers/merger/SimpleMerger/simple_merger.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+__copyright__ = "Copyright (c) 2021 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from collections import OrderedDict
+from typing import List, Tuple
+
+from jina import Executor, requests, DocumentArray
+
+
+class SimpleMerger(Executor):
```
Comment (JoanFM) [inline jinahub/indexers/merger/SimpleMerger/simple_merger.py:10]: The name should be `MatchMerger`. Or we need to pass what to `merge` , `chunks` or `matches`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/indexers/merger/SimpleMerger/tests/test_simple_merger.py (hunk 1 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/SimpleMerger/tests/test_simple_merger.py]: good point, I added unit tests

File: jinahub/indexers/merger/SimpleMerger/tests/unit/test_simple_merger.py (hunk 2 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/SimpleMerger/tests/unit/test_simple_merger.py]: This test was incomplete. Now I added some logic to it.

File: jinahub/indexers/merger/SimpleMerger/simple_merger.py (hunk 3 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/SimpleMerger/simple_merger.py]: thanks, I moved it

File: jinahub/indexers/merger/SimpleMerger/simple_merger.py (hunk 4 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/SimpleMerger/simple_merger.py:9]: Okay I will call it MatchMerger for now

File: jinahub/indexers/merger/MatchMerger/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,14 @@
+FROM jinaai/jina:2.0.18 as base
+# install and upgrade pip
+RUN apt-get update && apt-get install -y python3.7 python3.7-dev python3-pip git
+RUN python3.7 -m pip install --upgrade pip
```
Comment (tadejsv) [inline jinahub/indexers/merger/MatchMerger/Dockerfile:4]: Only installing git is needed

File: jinahub/indexers/merger/MatchMerger/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+## 🎉️ Example 
+
+```python
+from jina import Flow, Document
+
+f = Flow().add(uses='jinahub+docker://AnnoySearcher')
+
+with f:
+    resp = f.post(on='/search', inputs=Document(), return_results=True)
+    print(f'{resp}')
```
Comment (tadejsv) [inline jinahub/indexers/merger/MatchMerger/README.md:73]: Needs to be updated

File: jinahub/indexers/merger/MatchMerger/README.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+f = Flow().add(uses='jinahub://AnnoySearcher')
+```
+
+or in the `.yml` config.
+
+```yaml
+jtype: Flow
+pods:
+  - name: indexer
+    uses: 'jinahub://AnnoySearcher'
```
Comment (tadejsv) [inline jinahub/indexers/merger/MatchMerger/README.md:60]: Update (other examples as well)

File: jinahub/indexers/merger/MatchMerger/match_merger.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    def __init__(self, default_traversal_paths: Tuple[str, ...] = ('r',), **kwargs):
+        super().__init__(**kwargs)
+        self.default_traversal_paths = default_traversal_paths
+
+    @requests
+    def merge(self, docs_matrix: List[DocumentArray], parameters: dict, **kwargs):
+        traversal_paths = parameters.get('traversal_paths', self.default_traversal_paths)
+        results = {}
+        for docs in docs_matrix:
+            self._merge_shard(results, docs, traversal_paths)
```
Comment (tadejsv) [inline jinahub/indexers/merger/MatchMerger/match_merger.py:27]: What about some sorting of matches by score here?

**Review by tadejsv — CHANGES_REQUESTED** [review]

File: jinahub/indexers/merger/MatchMerger/match_merger.py (hunk 8 above)
Comment (JoanFM) [reply jinahub/indexers/merger/MatchMerger/match_merger.py:27]: If we add sorting remember to pass the top_k parameter to sort only up to that level. And pass the name of the score to be based on and wether is distance or similarity/relevance

File: jinahub/indexers/merger/MatchMerger/README.md (hunk 6 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/MatchMerger/README.md]: thanks, updated

File: jinahub/indexers/merger/MatchMerger/match_merger.py (hunk 8 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/MatchMerger/match_merger.py:27]: Nice idea.
I would like to not support top-k for now and also let the ranker sort the results.
Just to keep this PR lean.

File: jinahub/indexers/merger/MatchMerger/Dockerfile (hunk 5 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/MatchMerger/Dockerfile]: done

File: jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        for doc in resp.docs:
+            assert {d.tags['shard_id'] for d in doc.matches} == {float(i) for i in range(shards)}
+
+    with Flow().add(
+            uses=MockShard,
+            uses_after=MatchMerger,
+            shards=shards,
+            polling='all'
+    ) as f:
+        f.search(docs, on_done=callback)
```
Comment (JoanFM) [inline jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py:34]: we prefer testing it with return_results True

File: jinahub/indexers/storage/test.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import numpy as np
+
+from PostgreSQLStorage import PostgreSQLStorage
+from jina import DocumentArray, Document
+
+storage = PostgreSQLStorage(hostname="localhost", port=5432, username="postgresadmin", password="1235813", database="postgresdb", table="searcher2")
```
Comment (JoanFM) [inline jinahub/indexers/storage/test.py:6]: what is this?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py (hunk 9 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py]: check

File: jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py [hunk 11]
```diff
@@ -20,15 +20,13 @@ def docs():
...
     with Flow().add(
             uses=MockShard,
             uses_after=MatchMerger,
             shards=shards,
             polling='all'
     ) as f:
-        f.search(docs, on_done=callback)
+        resp = f.search(docs, return_results=True)[0].docs
+        assert len(resp) == 2
+        for doc in resp:
```
Comment (JoanFM) [inline jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py:31]: this does not work, resp is a list of responses

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jinahub/indexers/storage/test.py (hunk 10 above)
Comment (florian-hoenicke) [reply jinahub/indexers/storage/test.py]: sorry, this was just a test file

File: jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py (hunk 11 above)
Comment (florian-hoenicke) [reply jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py:31]: [0].docs

Comment (JoanFM) [reply jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py:31]: ah then do not call it resp if possibke

Comment (florian-hoenicke) [reply jinahub/indexers/merger/MatchMerger/tests/integration/test_match_merger.py]: check

**Review by JoanFM — DISMISSED** [review]

**Comment by cristianmtr** [comment]
Can you try to drop-in replace it in the integration tests,

here 
https://github.com/jina-ai/executors/blob/09d8235caac42e38b0ffca58dd22d00ae74d4b13/tests/integration/psql_dump_reload/test_dump_psql.py#L59
and here
https://github.com/jina-ai/executors/blob/98e9606d9c88802f94590d574d4efa8140684b5e/tests/integration/lmdb_dump_reload/test_dump_lmdb.py#L47

to get more integration tests where it is used

**Review by cristianmtr — CHANGES_REQUESTED** [review]
can u try adding it to the integration tests?

**Comment by florian-hoenicke** [comment]
> can u try adding it to the integration tests?
The match merger currently does not support top-k in the first iteration.
So, it can not act as replacement.

**Comment by cristianmtr** [comment]
> > can u try adding it to the integration tests?
> 
> The match merger currently does not support top-k in the first iteration.
> So, it can not act as replacement.

I think it should support top_k, right?

**Comment by florian-hoenicke** [comment]
> > > can u try adding it to the integration tests?
> > 
> > 
> > The match merger currently does not support top-k in the first iteration.
> > So, it can not act as replacement.
> 
> I think it should support top_k, right?

Yes, but maybe not in the first iteration.

**Comment by cristianmtr** [comment]
> > > > can u try adding it to the integration tests?
> > > 
> > > 
> > > The match merger currently does not support top-k in the first iteration.
> > > So, it can not act as replacement.
> > 
> > 
> > I think it should support top_k, right?
> 
> Yes, but maybe not in the first iteration.

can u open an issue to track that?

**Comment by florian-hoenicke** [comment]
> can u open an issue to track that?

Good point, let's make sure we don't forget about this.
https://github.com/jina-ai/executors/issues/113

File: jinahub/indexers/merger/MatchMerger/config.yml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+jtype: MatchMerger
+with:
+  {}
```
Comment (tadejsv) [inline jinahub/indexers/merger/MatchMerger/config.yml:3]:
```suggestion
```

**Review by tadejsv — DISMISSED** [review]
Looks good to me

File: jinahub/indexers/merger/MatchMerger/manifest.yml [hunk 13]
```diff
@@ -0,0 +1,9 @@
+manifest_version: 1
+name: MatchMerger
+alias: MatchMerger
```
Comment (nan-wang) [inline jinahub/indexers/merger/MatchMerger/manifest.yml:3]:
```suggestion
```

File: jinahub/indexers/merger/MatchMerger/manifest.yml [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+description: Merges the results of shards by appending all matches.
+author: Jina AI Dev-Team (dev-team@jina.ai)
```
Comment (nan-wang) [inline jinahub/indexers/merger/MatchMerger/manifest.yml:5]:
```suggestion
```

File: jinahub/indexers/merger/MatchMerger/manifest.yml [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+url: https://jina.ai
+documentation: https://github.com/jina-ai/jina-hub
```
Comment (nan-wang) [inline jinahub/indexers/merger/MatchMerger/manifest.yml:7]:
```suggestion
```

File: jinahub/indexers/merger/MatchMerger/manifest.yml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+manifest_version: 1
+name: MatchMerger
+alias: MatchMerger
+description: Merges the results of shards by appending all matches.
+author: Jina AI Dev-Team (dev-team@jina.ai)
+url: https://jina.ai
```
Comment (nan-wang) [inline jinahub/indexers/merger/MatchMerger/manifest.yml:6]:
```suggestion
url: https://github.com/jina-ai/executors/tree/main/jinahub/indexers/merger/MatchMerger
```

File: jinahub/indexers/merger/MatchMerger/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,10 @@
+FROM jinaai/jina:2.0.18 as base
+# install and upgrade pip
+RUN apt-get update && apt-get install -y git
```
Comment (nan-wang) [inline jinahub/indexers/merger/MatchMerger/Dockerfile:3]: Do we really need `git`?
```suggestion
```

**Review by nan-wang — CHANGES_REQUESTED** [review]
check out PR #110

**Review by tadejsv — APPROVED** [review]
LGTM

**Review by nan-wang — APPROVED** [review]
LGTM👍

**Pull request merged**